lightningd/options.c

```diff
@@ -366,6 +368,58 @@ static char *opt_clear_plugins(struct lightningd *ld)
 	return NULL;
 }
 
+/* Prompt the user to enter a password, from which will be derived the key used
+ * for `hsm_secret` encryption.
+ * The algorithm used to derive the key is Argon2(id), to which libsodium
+ * defaults. However argon2id-specific constants are used in case someone runs it
+ * with a libsodium version which default constants differs (typically <1.0.9).
+ */
+static char *opt_set_hsm_password(struct lightningd *ld)
+{
+	struct termios current_term, temp_term;
+	char *passwd = NULL;
+	size_t passwd_size = 0;
+	u8 salt[11] = "c-lightning";
+	ld->encrypted_hsm = true;
+
+	ld->config.keypass = tal(NULL, struct secret);
+	/* Don't swap the encryption key ! */
+	if (sodium_mlock(ld->config.keypass->data,
+	                 sizeof(ld->config.keypass->data)) != 0)
+		return "Could not lock hsm_secret encryption key memory.";
+
+	/* Get the password from stdin, but don't echo it. */
+	if (tcgetattr(fileno(stdin), &current_term) != 0)
+		return "Could not get current terminal options.";
+	temp_term = current_term;
+	temp_term.c_lflag &= ~ECHO;
+	if (tcsetattr(fileno(stdin), TCSAFLUSH, &temp_term) != 0)
+		return "Could not disable password echoing.";
+	printf("Enter hsm_secret password : ");
+	if (getline(&passwd, &passwd_size, stdin) < 0)
+		return "Could not read password from stdin.";
```

Comment:
I think we'd better ask users to enter the password twice and check the consistency.

Reply:
If two passwords are different, users must repeat this setting process.

Reply:
And if user enter `enter` or `null`, this process will be cancelled?

Reply:
The decryption is done in `hsmd`, and I think it's better to keep it that way so that only one daemon interacts with `hsm_secret`. In case of a wrong password, an user can still `arrow_up` then `enter` :-).

> And if user enter enter or null, this process will be cancelled?

No, actually entering `null` is a backdoor to crack any C-lightning node's master seed :wink: